### ISSUE TITLE
Document the Twisted dependancy list

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#11546]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)| @jyknight reported|
|-|-|
|Trac ID|trac#967|
|Type|enhancement|
|Created|2005-04-06 00:34:42Z|
|Branch|https://github.com/twisted/twisted/tree/dependancy-list-967-2|

<details><summary>Searchable metadata</summary>

```
trac-id__967 967
type__enhancement enhancement
reporter__jknight jknight
priority__low low
milestone__ 
branch__branches_dependancy_list_967_2 branches/dependancy-list-967-2
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__documentation documentation
time__1112747682000000 1112747682000000
changetime__1300247024000000 1300247024000000
version__ 
owner__exarkun exarkun
cc__radix cc__jknight cc__thijs cc__binjured cc__jesstess
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

```
#!html
<pre>
I can't find a list of what each twisted module depends on on our website. 
If it's there it must be well hidden.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="jyknight's avatar" src="https://avatars.githubusercontent.com/u/873903?s=50" width="50" height="50">](https://github.com/jyknight)|@jyknight commented|
|-|-|

```
#!html
<pre>
Tv mentions (in svn) /Twisted/debian/projects/*/control
and http://twistedmatrix.com/~tv/2.0-BEWARE/build-depends.dot.png
which show some of the info.
</pre>
```

**Comment by twisted-trac** [comment]
|[<img alt="puzzlement's avatar" src="https://avatars.githubusercontent.com/u/94942?s=50" width="50" height="50">](https://github.com/puzzlement)|@puzzlement set owner to edsuom|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_4"></a>|@thijstriemstra set owner to @thijstriemstra<br>@thijstriemstra set status to `assigned`|
|-|-|

The pylint tool generates a dependency tree that we could add on the wiki or in a textfile, updating it for every release wouldn't be too hard and it also could be automated. It's better than nothing.. Current one for web looks like the page below. I'd suggest removing the references on a module level (in parentheses).

```
#!html
<h2>External dependencies</h2>
<pre>zope 
  \-interface (web.http,web.iweb,web.monitor,web.resource,web._auth.basic,web._auth.digest,web._auth.wrapper,web.test.test_httpauth,twisted.web.test.test_web,twisted.web.woven.controller,twisted.web.woven.interfaces,web.woven.model,web.woven.simpleguard,web.woven.template,web.woven.utils,twisted.web.woven.view)
    \-verify (web.test.test_httpauth)
twisted 
  \-copyright (web.script,web.server)
  \-trial 
  | \-unittest (web.test.test_cgi,web.test.test_distrib,web.test.test_domhelpers,web.test.test_http,web.test.test_http_headers,web.test.test_httpauth,web.test.test_mvc,web.test.test_proxy,web.test.test_soap,web.test.test_static,web.test.test_tap,twisted.web.test.test_web,web.test.test_webclient,web.test.test_woven,web.test.test_xml,web.test.test_xmlrpc)
  \-python 
  | \-compat (web.test.test_http,web.test.test_http_headers)
  | \-reflect (web.server,web.sux,web.tap,web.widgets,web.xmlrpc,twisted.web.woven.input,web.woven.model,twisted.web.woven.page,web.woven.tapestry,twisted.web.woven.widgets)
  | \-runtime (web.static)
  | \-log (web.demo,web.distrib,web.html,web.http,web.server,web.twcgi,web.widgets,web.xmlrpc,web._auth.wrapper,web.test.test_distrib,twisted.web.test.test_web,twisted.web.woven.controller,web.woven.flashconduit,web.woven.guard,twisted.web.woven.input,web.woven.template,web.woven.utils,twisted.web.woven.view,twisted.web.woven.widgets)
  | \-filepath (web.static,web.twcgi,web.test.test_static,web.test.test_webclient,web.woven.dirlist)
  | \-urlpath (web.server,web.util)
  | \-rebuild (web.widgets)
  | \-versions (web._version)
  | \-util (web.client,web.microdom,web.static,web.widgets,web.test.test_cgi)
  | \-failure (web.client,web.server,web.util,web.widgets,web.xmlrpc,web.test.test_xmlrpc,twisted.web.woven.controller,web.woven.form,web.woven.template,web.woven.utils,twisted.web.woven.widgets)
  | \-formmethod (web.woven.form,web.woven.guard)
  | \-components (web.server,web.static,web.widgets,web._auth.wrapper,web.test.test_mvc,twisted.web.woven.controller,web.woven.form,web.woven.guard,web.woven.model,web.woven.template,web.woven.utils,twisted.web.woven.view,twisted.web.woven.widgets)
  | \-randbytes (web._auth.digest)
  | \-usage (web.tap)
  | \-logfile (web.server)
  | \-roots (web.vhost)
  \-application 
  | \-strports (web.tap)
  | \-service (web.tap)
  | \-internet (web.tap)
  \-persisted 
  | \-styles (web.distrib,web.static)
  \-spread 
  | \-pb (web.distrib,web.server,web.static,web.tap,web.twcgi,web.test.test_distrib,web.test.test_tap)
  \-internet 
  | \-defer (web.client,web.google,web.server,web.soap,web.widgets,web.xmlrpc,web.test.test_distrib,web.test.test_soap,twisted.web.test.test_web,web.test.test_webclient,web.test.test_woven,web.test.test_xmlrpc,web.woven.form,twisted.web.woven.input,web.woven.model,web.woven.tapestry,web.woven.template,twisted.web.woven.view,twisted.web.woven.widgets)
  | \-task (web.server,twisted.web.test.test_web)
  | \-protocol (web.client,web.google,web.http,web.proxy,web.sux,web.twcgi,web.xmlrpc,web.test.test_http,web.woven.flashconduit)
  | \-reactor (web.client,web.distrib,web.google,web.http,web.monitor,web.proxy,web.twcgi,web.xmlrpc,web.test.test_cgi,web.test.test_distrib,web.test.test_soap,web.test.test_webclient,web.test.test_woven,web.test.test_xmlrpc,web.woven.guard,web.woven.template)
  | \-interfaces (web.http,web.tap,web.test.test_cgi,twisted.web.test.test_web,web.test.test_webclient)
  | \-address (web.distrib,web.http,web.server,web.test.test_httpauth,twisted.web.test.test_web,web.test.test_woven)
  | \-ssl (web.client,web.tap,web.xmlrpc,web.test.test_webclient)
  | \-error (web.test.test_proxy,twisted.web.test.test_web,web.test.test_xmlrpc)
  | \-abstract (web.static)
  \-test (web.test.test_webclient)
  | \-proto_helpers (web.test.test_http,web.test.test_proxy)
  | \-test_protocols (web.test.test_http)
  \-cred 
  | \-credentials (web._auth.basic,web._auth.digest,web.test.test_httpauth,web.woven.guard)
  | \-checkers (web.test.test_httpauth,web.woven.simpleguard)
  | \-portal (web.test.test_httpauth,web.woven.simpleguard)
  | \-error (web._auth.basic,web._auth.digest,web._auth.wrapper,web.test.test_httpauth,web.woven.guard)
  \-protocols (web.http,web.server,web.test.test_http)
    \-basic (web.http,web.woven.flashconduit)
    \-policies (web.http,web.test.test_webclient)
    \-loopback (web.test.test_http)
SOAPpy (web.soap,web.test.test_soap)</pre></div>
<div>
```

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_5"></a>|@thijstriemstra removed owner<br>@thijstriemstra set status to `new`|
|-|-|

Actually setuptools provides a feature for this: ```extras_require```. It's a place to list dependencies (that are available to pypi). Thoughts? (I'll need the list of dependencies before I can work on this)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set owner to @thijstriemstra|
|-|-|

I don't know who would benefit from the information above generated by pylint.  The only parts of it that are interesting are the zope.interface part and the SOAPpy part.  I would suggest adding this information to the README files.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_7"></a>|@thijstriemstra commented|
|-|-|

(In [26438]) Branching to 'dependancy-list-967'

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_8"></a>|@thijstriemstra removed owner|
|-|-|

Added extras_require to setup.py for listing Twisted optional dependancies in r26442. For now only pycrypto is in there, but it should list all of Twisted's optional depedancies. If somebody can give me the full list I'll add all of them there.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_9"></a>|@thijstriemstra commented|
|-|-|

Description of extras_require: http://peak.telecommunity.com/DevCenter/setuptools#declaring-extras-optional-features-with-their-own-dependencies

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_10"></a>|@glyph set owner to @thijstriemstra|
|-|-|

It seems like this functionality should really be in `twisted.python.dist` or `twisted.python.release` or something, especially given that `twisted/conch/topfiles/setup.py` needs to know about the pycrypto dependency somehow too.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun commented|
|-|-|

Isn't this a documentation ticket?  Changing setup.py files doesn't seem like the way to resolve a documentation ticket.

**Comment by twisted-trac** [comment]
|[<img alt="thijstriemstra's avatar" src="https://avatars.githubusercontent.com/u/305679?s=50" width="50" height="50">](https://github.com/thijstriemstra)<a name="note_12"></a>|@thijstriemstra commented|
|-|-|

Replying to [glyph](#note_10):
> It seems like this functionality should really be in `twisted.python.dist` or `twisted.python.release` or something, especially given that `twisted/conch/topfiles/setup.py` needs to know about the pycrypto dependency somehow too.

I opened [[[#3696](https://github.com/twisted/twisted/issues/3696)](https://github.com/twisted/twisted/issues/3696)](https://github.com/twisted/twisted/issues/3696) for this.

Replying to [exarkun](#note_11):
> Isn't this a documentation ticket?  Changing setup.py files doesn't seem like the way to resolve a documentation ticket.

What would be the best approach to add documentation for those extra_requires dependencies.. manually, by adding it to the README like [twisted/conch/topfiles/README conch](twisted/conch/topfiles/README conch) does for example, or hack something into #3696? I think it's a good idea to both document this and also programmatically declare those dependencies using setuptools extras_require so it can be used for discovering Twisted's dependencies in other programs using Twisted.. So I think this one should serve as the doc ticket and #3696 for the coding part.

**Comment by twisted-trac** [comment]
|<img alt="binjured's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_14"></a>|binjured commented|
|-|-|

Anyone care to chime in and give some guidance on this issue?

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_15"></a>|@exarkun commented|
|-|-|

The last thing thijs said makes sense to me.  This ticket should be for updating READMEs.  [#3696](https://github.com/twisted/twisted/issues/3696) can be for doing things to the release tool chain.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_17"></a>|@exarkun commented|
|-|-|

(In [30939]) Branching to 'dependancy-list-967-2'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_18"></a>|@exarkun commented|
|-|-|

(In [30940]) Update dependencies for various subprojects

refs [#11546](https://github.com/twisted/twisted/issues/11546)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_19"></a>|@exarkun commented|
|-|-|

I changed some READMEs.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)|@jesstess set owner to @jesstess|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_22"></a>|@jesstess set owner to @exarkun|
|-|-|

Who doesn't love 3 digit Twisted tickets? Thanks for working on this, exarkun. Some feedback:

 * INSTALL: There are nicer URL options for pywin32 now: maybe http://sourceforge.net/projects/pywin32/ or http://sourceforge.net/projects/pywin32/files/
 * Twisted Mail: ```twisted.mail.protocols.SSLContextFactory``` uses ```OpenSSL```
 * Twisted Conch:
  * depends on Twisted Core.
  * ```twisted/conch/scripts/tkconch.py``` and ```twisted/conch/ui/tkvt100.py``` use Tkinter.
 * Twisted Web:
  * "Twisted Web depends on Twisted" ==> "Twisted Web depends on Twisted Core".
  * ```twisted/web/script.py``` has a ```quixote``` import in ```ResourceTemplate```. Maybe ```ResourceTemplate``` was supposed to die, but it is referenced in a few howtos.
 * Twisted Pair. Since it's unmaintained, I don't know if we care about updating the README, but:
  * Twisted Pair depends on Twisted Core.
  * ```twisted/pair/tuntap.py``` has a long docstring detailing its dependency on ```python-eunuchs```, which doesn't  seem to exist anymore.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_23"></a>|@exarkun commented|
|-|-|

(In [30988]) Add more dependency information, update links to point at more friendly pages

refs [#11546](https://github.com/twisted/twisted/issues/11546)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_24"></a>|@exarkun removed owner|
|-|-|

Thanks for the review!  I think the revision in the previous comment addresses all of those comments.

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)|@jesstess set owner to @jesstess|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="jesstess's avatar" src="https://avatars.githubusercontent.com/u/188336?s=50" width="50" height="50">](https://github.com/jesstess)<a name="note_26"></a>|@jesstess set owner to @exarkun|
|-|-|

Looks great, please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_27"></a>|@exarkun set status to `closed`|
|-|-|

(In [31170]) Merge dependancy-list-967-2

Author: exarkun
Reviewer: jesstess
Fixes: [#11546](https://github.com/twisted/twisted/issues/11546)

Add the dependencies of each project to the README file of each subproject
and the project-wide INSTALL file.
